datafusion/core/src/datasource/physical_plan/parquet/row_filter.rs

```diff
@@ -268,90 +261,192 @@ impl<'a> FilterCandidateBuilder<'a> {
     /// * `Ok(Some(candidate))` if the expression can be used as an ArrowFilter
     /// * `Ok(None)` if the expression cannot be used as an ArrowFilter
     /// * `Err(e)` if an error occurs while building the candidate
-    pub fn build(
-        mut self,
-        metadata: &ParquetMetaData,
-    ) -> Result<Option<FilterCandidate>> {
-        let expr = self.expr.clone().rewrite(&mut self).data()?;
-
-        if self.non_primitive_columns || self.projected_columns {
-            Ok(None)
-        } else {
-            let required_bytes =
-                size_of_columns(&self.required_column_indices, metadata)?;
-            let can_use_index = columns_sorted(&self.required_column_indices, metadata)?;
-
-            Ok(Some(FilterCandidate {
-                expr,
-                required_bytes,
-                can_use_index,
-                projection: self.required_column_indices.into_iter().collect(),
-            }))
+    pub fn build(self, metadata: &ParquetMetaData) -> Result<Option<FilterCandidate>> {
+        let Some((projection, rewritten_expr)) = pushdown_columns(
+            Arc::clone(&self.expr),
+            self.file_schema,
+            self.table_schema,
+        )?
+        else {
+            return Ok(None);
+        };
+
+        let required_bytes = size_of_columns(&self.required_column_indices, metadata)?;
+        let can_use_index = columns_sorted(&self.required_column_indices, metadata)?;
+
+        Ok(Some(FilterCandidate {
+            expr: rewritten_expr,
+            required_bytes,
+            can_use_index,
+            projection,
+        }))
+    }
+}
+
+// a struct that implements TreeNodeRewriter to traverse a PhysicalExpr tree structure to determine
```

Comment:
Thank you veru much for adding additional comments -- it is a great example of leaving the code better than you found it